src/NXP/Classes/CustomFunction.php

```diff
@@ -32,28 +33,29 @@ class CustomFunction
      * @throws ReflectionException
      * @throws IncorrectNumberOfFunctionParametersException
      */
-    public function __construct(string $name, callable $function, ?int $places = null)
-    {
+    public function __construct(string $name, callable $function, ?int $places = null) {
         $this->name = $name;
         $this->function = $function;
         if ($places === null) {
             $reflection = new ReflectionFunction($function);
-            $this->places = $reflection->getNumberOfParameters();
+            $this->places = $reflection->getNumberOfRequiredParameters();
+            $this->placesOptional = $reflection->getNumberOfParameters();
         } else {
             $this->places = $places;
+            $this->placesOptional = $places;
         }
     }
 
-    public function execute(array &$stack) : Token
-    {
+    public function execute(array &$stack): Token {
         if (count($stack) < $this->places) {
             throw new IncorrectNumberOfFunctionParametersException($this->name);
         }
         $args = [];
-        for ($i = 0; $i < $this->places; $i++) {
+
+        for ($i = 0; $i < $this->placesOptional; $i++) {
             array_unshift($args, array_pop($stack)->value);
         }
-
+        $args = array_values(array_filter($args));
```

Comment:
This line breaks tests. I suggest pass callback that strictly checks argument to null